sw/host/opentitanlib/src/app/ioexpander.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::app::config;
+use crate::app::{PinStrapping, TransportWrapper};
+use crate::io::gpio::{GpioError, GpioPin, PinMode, PullMode};
+use crate::io::i2c::{self, Bus};
+use crate::transport::TransportError;
+
+use anyhow::{bail, Result};
+use std::rc::Rc;
+use std::vec::Vec;
+
+pub struct IoExpander {
+    pub pins: Vec<Rc<dyn GpioPin>>,
+}
+
+struct Sx1503 {
+    mux_strapping: Option<PinStrapping>,
+    i2c_bus: Rc<dyn Bus>,
+    i2c_addr: u8,
+}
+
+#[allow(dead_code)]
+#[derive(Clone, Copy)]
+enum Sx1503Registers {
+    DataB = 0x00,
+    DataA = 0x01,
+    DirB = 0x02,
+    DirA = 0x03,
+    PullUpB = 0x04,
+    PullUpA = 0x05,
+    PullDownB = 0x06,
+    PullDownA = 0x07,
+    InterruptMaskB = 0x08,
+    InterruptMaskA = 0x09,
+    SenseHighB = 0x0A,
+    SenseHighA = 0x0B,
+    SenseLowB = 0x0C,
+    SenseLowA = 0x0D,
+    InterruptSourceB = 0x0E,
+    InterruptSourceA = 0x0F,
+    EventStatusB = 0x10,
+    EventStatusA = 0x11,
+    PLDModeB = 0x20,
+    PLDModeA = 0x21,
+    PLDTable0B = 0x22,
+    PLDTable0A = 0x23,
+    PLDTable1B = 0x24,
+    PLDTable1A = 0x25,
+    PLDTable2B = 0x26,
+    PLDTable2A = 0x27,
+    PLDTable3B = 0x28,
+    PLDTable3A = 0x29,
+    PLDTable4B = 0x2A,
+    PLDTable4A = 0x2B,
+    Advanced = 0xAD,
+}
+
+impl Sx1503 {
+    fn apply_mux_strapping(&self) -> Result<()> {
+        if let Some(ref strapping) = self.mux_strapping {
+            strapping.apply()?
+        }
+        Ok(())
+    }
+
+    fn remove_mux_strapping(&self) -> Result<()> {
+        if let Some(ref strapping) = self.mux_strapping {
+            strapping.remove()?
+        }
+        Ok(())
+    }
+
+    fn read_register(&self, addr: Sx1503Registers) -> Result<u8> {
+        let mut val: u8 = 0;
+        self.i2c_bus.run_transaction(
+            self.i2c_addr,
+            &mut [
+                i2c::Transfer::Write(&[addr as u8]),
+                i2c::Transfer::Read(std::slice::from_mut(&mut val)),
+            ],
+        )?;
+        Ok(val)
+    }
+
+    fn write_register(&self, addr: Sx1503Registers, data: u8) -> Result<()> {
+        self.i2c_bus.run_transaction(
+            self.i2c_addr,
+            &mut [i2c::Transfer::Write(&[addr as u8, data])],
+        )?;
+        Ok(())
+    }
+
+    fn set_or_clear_bit(&self, addr: Sx1503Registers, bit_no: u8, value: bool) -> Result<()> {
+        let val = self.read_register(addr)?;
+        let val = val & !(1 << bit_no) | (if value { 1 << bit_no } else { 0 });
+        self.write_register(addr, val)?;
+        Ok(())
+    }
+}
+
+pub fn new_sx1503(
```

Comment:
Should this be an associated function of `Sx1503`?  e.g:
```
impl Sx1503 {
    pub fn new(...) -> Result<...> { }
}

Reply:
It could, but I am actually happy now that struct `Sx1503` is not public from the `sx1503.rs` file, only the single `new()` method (now called `create()`) is public, and even that is visible only to the dispatching `create()` method in `src/transport/ioexpander/mod.rs`, not to `TransportWrapper`.